File: packages/wos-plugin-sdk-python/src/wos_plugin_sdk/logger.py
"""
WorldOS Plugin SDK - Logger

Structured logging for plugins with configurable log levels.
"""

import json
import os
import sys
from datetime import datetime, timezone
from enum import IntEnum
from typing import Any, Optional, TextIO


class LogLevel(IntEnum):
    """Log levels for the plugin logger."""

    DEBUG = 0
    INFO = 1
    WARN = 2
    ERROR = 3
    SILENT = 4


# Global log level
_global_log_level: LogLevel = LogLevel.INFO


def set_global_log_level(level: LogLevel) -> None:
    """Set the global log level for all loggers."""
    global _global_log_level
    _global_log_level = level


def get_global_log_level() -> LogLevel:
    """Get the current global log level."""
    return _global_log_level


class Logger:
    """
    Structured logger for WorldOS plugins.

    Outputs JSON-formatted log messages to stdout/stderr
    with timestamps and log levels.
    """

    def __init__(
        self,
        name: str,
        level: Optional[LogLevel] = None,
        output: Optional[TextIO] = None,
    ) -> None:
        """
        Create a new logger.

        Args:
            name: Logger name (usually plugin name)
            level: Log level override (defaults to global level)
            output: Output stream (defaults to stdout for info/debug, stderr for warn/error)
        """
        self.name = name
        self._level = level
        self._output = output

    @property
    def level(self) -> LogLevel:
        """Get the effective log level."""
        return self._level if self._level is not None else _global_log_level

    def _format_message(
        self,
        level: str,
        message: str,
        data: Optional[dict[str, Any]] = None,
    ) -> str:
        """Format a log message as JSON."""
        log_entry = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "level": level,
            "logger": self.name,
            "message": message,
        }
        if data:
            log_entry["data"] = data
        return json.dumps(log_entry)

    def _output_stream(self, level: LogLevel) -> TextIO:
        """Get the output stream for a log level."""
        if self._output:
            return self._output
        return sys.stderr if level >= LogLevel.WARN else sys.stdout

    def _log(
        self,
        level: LogLevel,
        level_name: str,
        message: str,
        data: Optional[dict[str, Any]] = None,
    ) -> None:
        """Internal log method."""
        if level < self.level:
            return
        formatted = self._format_message(level_name, message, data)
        output = self._output_stream(level)
        print(formatted, file=output)

    def debug(self, message: str, data: Optional[dict[str, Any]] = None) -> None:
        """Log a debug message."""
        self._log(LogLevel.DEBUG, "debug", message, data)

    def info(self, message: str, data: Optional[dict[str, Any]] = None) -> None:
        """Log an info message."""
        self._log(LogLevel.INFO, "info", message, data)

    def warn(self, message: str, data: Optional[dict[str, Any]] = None) -> None:
        """Log a warning message."""
        self._log(LogLevel.WARN, "warn", message, data)

    def error(
        self,
        message: str,
        error: Optional[Exception] = None,
        data: Optional[dict[str, Any]] = None,
    ) -> None:
        """Log an error message."""
        if error:
            data = data or {}
            data["error"] = {
                "type": type(error).__name__,
                "message": str(error),
            }
        self._log(LogLevel.ERROR, "error", message, data)

    def child(self, name: str) -> "Logger":
        """Create a child logger with an extended name."""
        return Logger(f"{self.name}:{name}", self._level, self._output)


def create_logger(
    name: Optional[str] = None,
    level: Optional[LogLevel] = None,
) -> Logger:
    """
    Create a logger instance.

    If no name is provided, uses the WOS_PLUGIN_NAME environment variable.

    Args:
        name: Logger name (defaults to WOS_PLUGIN_NAME env var)
        level: Log level (defaults to global level)

    Returns:
        Logger instance
    """
    if name is None:
        name = os.environ.get("WOS_PLUGIN_NAME", "plugin")
    return Logger(name, level)

File: packages/wos-plugin-sdk/src/logging/logger.ts
/**
 * SDK Logging Integration
 *
 * Story 2.7: SDK Logging Integration
 *
 * Provides structured logging with configurable levels for plugins.
 */

/**
 * Log levels in order of severity
 */
export type LogLevel = 'debug' | 'info' | 'warn' | 'error' | 'silent';

/**
 * Logger interface
 */
export interface Logger {
  debug(message: string, meta?: Record<string, unknown>): void;
  info(message: string, meta?: Record<string, unknown>): void;
  warn(message: string, meta?: Record<string, unknown>): void;
  error(message: string, meta?: Record<string, unknown>): void;
  child(name: string): Logger;
  setLevel(level: LogLevel): void;
}

/**
 * Options for creating a logger
 */
export interface LoggerOptions {
  /** Log level for this logger (overrides global) */
  level?: LogLevel;
}

/**
 * Log level priority (higher = more severe)
 */
const LOG_LEVEL_PRIORITY: Record<LogLevel, number> = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3,
  silent: 4,
};

/**
 * Global log level (can be overridden per-logger)
 */
let globalLogLevel: LogLevel = 'info';

/**
 * Set the global log level
 */
export function setGlobalLogLevel(level: LogLevel): void {
  globalLogLevel = level;
}

/**
 * Get the current global log level
 */
export function getGlobalLogLevel(): LogLevel {
  return globalLogLevel;
}

/**
 * Format metadata for log output
 */
function formatMeta(meta?: Record<string, unknown>): string {
  if (!meta || Object.keys(meta).length === 0) {
    return '';
  }

  try {
    // Handle Error objects specially
    const processed: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(meta)) {
      if (value instanceof Error) {
        processed[key] = {
          name: value.name,
          message: value.message,
          stack: value.stack,
        };
      } else {
        processed[key] = value;
      }
    }
    return ' ' + JSON.stringify(processed);
  } catch {
    return ' [meta serialization failed]';
  }
}

/**
 * Create a logger for a plugin
 *
 * @param pluginName - Name of the plugin (used as prefix)
 * @param options - Logger options
 * @returns Logger instance
 */
export function createLogger(
  pluginName: string,
  options: LoggerOptions = {},
): Logger {
  let logLevel = options.level;

  const shouldLog = (level: LogLevel): boolean => {
    const effectiveLevel = logLevel ?? globalLogLevel;
    return LOG_LEVEL_PRIORITY[level] >= LOG_LEVEL_PRIORITY[effectiveLevel];
  };

  const formatMessage = (
    level: LogLevel,
    message: string,
    meta?: Record<string, unknown>,
  ): string => {
    const timestamp = new Date().toISOString();
    const metaStr = formatMeta(meta);
    return `${timestamp} ${level.toUpperCase()} [${pluginName}] ${message}${metaStr}`;
  };

  const logger: Logger = {
    debug(message: string, meta?: Record<string, unknown>): void {
      if (shouldLog('debug')) {
        console.debug(formatMessage('debug', message, meta));
      }
    },

    info(message: string, meta?: Record<string, unknown>): void {
      if (shouldLog('info')) {
        console.info(formatMessage('info', message, meta));
      }
    },

    warn(message: string, meta?: Record<string, unknown>): void {
      if (shouldLog('warn')) {
        console.warn(formatMessage('warn', message, meta));
      }
    },

    error(message: string, meta?: Record<string, unknown>): void {
      if (shouldLog('error')) {
        console.error(formatMessage('error', message, meta));
      }
    },

    child(name: string): Logger {
      return createLogger(`${pluginName}:${name}`, { level: logLevel });
    },

    setLevel(level: LogLevel): void {
      logLevel = level;
    },
  };

  return logger;
}
